### PR TITLE
Fix deadlock on Factory

### PR DESCRIPTION
**Pull request by 304NotModified**
- No locking on factory when rendering layout renderers
- Less locking when using GlobalDiagnosticsContext.Get / MappedDiagnosticsContext.Get / MappedDiagnosticsLogicalContext.Get
- added <remarks> for (indirect) locks 
- move ConvertToString from GlobalContext to FormatHelper (internal)
- introduced StringBuilderExt, remove duplicate code (internal)

fixes https://github.com/NLog/NLog/issues/1203

### CODE REVIEW

File: src/NLog/LayoutRenderers/MdlcLayoutRenderer.cs [hunk 1]
```diff
@@ -58,8 +59,14 @@ public class MdlcLayoutRenderer : LayoutRenderer
...
-            var message = MappedDiagnosticsLogicalContext.Get(Item);
-            builder.Append(message);
+            //don't use MappedDiagnosticsLogicalContext.Get to ensure we are not locking the Factory (indirect by LogManager.Configuration).
+            var o = MappedDiagnosticsLogicalContext.GetObject(this.Item);
+            var formatProvider = logEvent.FormatProvider;
+            if (formatProvider == null && LoggingConfiguration != null)
+            {
+                formatProvider = LoggingConfiguration.DefaultCultureInfo;
+            }
+            builder.Append(Convert.ToString(o, formatProvider));
```
Comment (Page-Not-Found) [inline src/NLog/LayoutRenderers/MdlcLayoutRenderer.cs]: Wouldn't it benefit if we add this logic to the base class and have it is a property?

Comment (304NotModified) [inline src/NLog/LayoutRenderers/MdlcLayoutRenderer.cs]: That's a good idea :)

**Comment by 304NotModified** [comment]
@Page-Not-Found I removed the code duplication.

**Comment by Page-Not-Found** [comment]
I did not find any other issues with this change.

**Comment by 304NotModified** [comment]
Thanks @Page-Not-Found !

**Pull request merged**